planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
+* Re-evaluate the `ViewProjection` fluent interface. That might fit in fine to our new model of projections for v4 with more overloads
+* I've been in favor of having adapters for existing projection libraries like [Liquid Projections](https://liquidprojections.net/) in the past, but I think we'd be losing too
+  much optization, especially for the aggregations inside the async daemon.
+
+#### Defining a Projection to a Document
+
+
+```
+ProjectTo<ProjectedDocument>()
+    // How to transform the events to the projected document, see below
+    .TransformWith<SampleTransformer>()
+
+    // Optionally aggregate by stream
+    .AggregateByStream() // Think this is self-descriptive
+
+    .AggregateByTenant() // Not sure how common this would be. Aggregate across the entire tenant
+
+    // Or aggregate by something special, hopefull use a common base type here for the event type
+    .AggregateByEventData<TEvent>(Expression<Func<TEvent, int>> expression) // what's the identity of the projected, aggregate doc?
+
+    .DeleteOn<TEvent>() // the aggregate document should be deleted if an event of this type hits
+                        // In this case, we could look ahead and "know" in the rebuilds not to rebuild
+                        // a projection for this aggregate as an optimization
+
+    .SnapshotInIncrementsOf(10) // Then maybe there are other options to retain old snapshots or not? Dunno
+    
+```
+
+
+Now that folks are used to the convention based signature approach of `StartUp` in ASP.Net Core, I suggest doing something
+similar for the projections like so:
+
+```
+    // This purposely does not implement any kind of interface
+    // Defines how to apply changes to one kind of projected document
+    public class SampleTransformer
```

Comment:
It would be good to see the example usage. I'm not sure if I get the idea behind that class.

Reply:
It's mostly to not have to have a bazillion overloads on the fluent interface to define projection steps. I'm not actually a fan of `ViewProjection` or anything that makes you code by inline Lambdas. Gives you more flexibility to use sync/async signatures, to use more services from the underlying container, to switch between the raw event, event wrappers, and stream information.

Reply:
Ok, that seems legit.